subprojects/base-services/src/main/java/org/gradle/util/CollectionUtils.java

```diff
@@ -184,20 +184,22 @@
         return destination;
     }
 
-    public static <R, I> List<R> collect(List<? extends I> list, Transformer<? extends R, ? super I> transformer) {
-        return collect(list, new ArrayList<R>(list.size()), transformer);
+    public static <R, I> Set<R> collect(Set<? extends I> set, Transformer<? extends R, ? super I> transformer) {
+        return collect(set, new HashSet<R>(set.size()), transformer);
     }
 
     public static <R, I> List<R> collect(I[] list, Transformer<? extends R, ? super I> transformer) {
         return collect(Arrays.asList(list), transformer);
     }
 
-    public static <R, I> Set<R> collect(Set<? extends I> set, Transformer<? extends R, ? super I> transformer) {
-        return collect(set, new HashSet<R>(), transformer);
-    }
-
     public static <R, I> List<R> collect(Iterable<? extends I> source, Transformer<? extends R, ? super I> transformer) {
-        return collect(source, new LinkedList<R>(), transformer);
+        if (source instanceof Collection<?>) {
+            //noinspection unchecked
+            int size = ((Collection<? extends I>) source).size();
```

Comment:
We have `Cast.unchecked()` and `Cast.uncheckedNonnull()` for these usages.

Reply:
Thanks, I'll extract it to a local variable and use it.